pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -242,64 +256,132 @@ func (m *Manager) ShutdownStatus() error {
-				gracePeriodOverride = *pod.Spec.TerminationGracePeriodSeconds
-			}
+				// If the pod's spec specifies a termination gracePeriod which is less than the gracePeriodOverride calculated, use the pod spec termination gracePeriod.
+				if pod.Spec.TerminationGracePeriodSeconds != nil && *pod.Spec.TerminationGracePeriodSeconds <= gracePeriodOverride {
+					gracePeriodOverride = *pod.Spec.TerminationGracePeriodSeconds
+				}
 
-			klog.V(1).InfoS("Shutdown manager killing pod with gracePeriod", "pod", klog.KObj(pod), "gracePeriod", gracePeriodOverride)
+				klog.V(1).InfoS("Shutdown manager killing pod with gracePeriod", "pod", klog.KObj(pod), "gracePeriod", gracePeriodOverride)
 
-			status := v1.PodStatus{
-				Phase:   v1.PodFailed,
-				Reason:  nodeShutdownReason,
-				Message: nodeShutdownMessage,
-			}
+				status := v1.PodStatus{
+					Phase:   v1.PodFailed,
+					Reason:  nodeShutdownReason,
+					Message: nodeShutdownMessage,
+				}
 
-			err := m.killPod(pod, status, &gracePeriodOverride)
-			if err != nil {
-				klog.V(1).InfoS("Shutdown manager failed killing pod", "pod", klog.KObj(pod), "err", err)
-			} else {
-				klog.V(1).InfoS("Shutdown manager finished killing pod", "pod", klog.KObj(pod))
-			}
-		}(pod)
-	}
+				err := m.killPod(pod, status, &gracePeriodOverride)
+				if err != nil {
+					klog.V(1).InfoS("Shutdown manager failed killing pod", "pod", klog.KObj(pod), "err", err)
+				} else {
+					klog.V(1).InfoS("Shutdown manager finished killing pod", "pod", klog.KObj(pod))
+				}
+			}(pod, group)
+		}
 
-	c := make(chan struct{})
-	go func() {
-		defer close(c)
-		wg.Wait()
-	}()
+		c := make(chan struct{})
+		go func() {
+			defer close(c)
+			wg.Wait()
+		}()
 
-	// We want to ensure that inhibitLock is released, so only wait up to the shutdownGracePeriodRequested timeout.
-	select {
-	case <-c:
-		break
-	case <-time.After(m.shutdownGracePeriodRequested):
-		klog.V(1).InfoS("Shutdown manager pod killing time out", "gracePeriod", m.shutdownGracePeriodRequested)
+		select {
+		case <-c:
+		case <-time.After(time.Duration(group.ShutdownGracePeriodSeconds) * time.Second):
+			klog.V(1).InfoS("Shutdown manager pod killing time out", "gracePeriod", group.ShutdownGracePeriodSeconds, "priority", group.Priority)
+		}
 	}
 
 	m.dbusCon.ReleaseInhibitLock(m.inhibitLock)
 	klog.V(1).InfoS("Shutdown manager completed processing shutdown event, node will shutdown shortly")
 
 	return nil
 }
+
+func (m *managerImpl) periodRequested() time.Duration {
+	var sum int64
+	for _, period := range m.podPriorityShutdownGracePeriod {
+		sum += period.ShutdownGracePeriodSeconds
+	}
+	return time.Duration(sum) * time.Second
+}
+
+func migrateConfig(shutdownGracePeriodRequested, shutdownGracePeriodCriticalPods time.Duration) []kubeletconfig.PodPriorityShutdownGracePeriod {
+	if shutdownGracePeriodRequested == 0 {
+		return nil
+	}
+	defaultPriority := shutdownGracePeriodRequested - shutdownGracePeriodCriticalPods
```

Comment:
Is it possible to have negative values here?  I am not sure the 0 check alone on GracePeriodRequested is sufficient.

Reply:
I think I would still feel better knowing there is a test in here that clips default and critical priority from going negative.

Reply:
Kubelet configuration validation has the check that ensures 

```
kc.ShutdownGracePeriodCriticalPods.Duration <= kc.ShutdownGracePeriod.Duration
```


https://github.com/kubernetes/kubernetes/blob/e67979eaf6e3ae9b8be67ae8374a3e008f88db76/pkg/kubelet/apis/config/validation/validation.go#L145-L147